### PR TITLE
fix(core): short circuit test executor when no tests and allowEmpty

### LINKED ISSUE #4475: `allowEmpty = true` exits with error if `thresholds` is set
**Summary**

StrykerJS exits with error code 1 when `allowEmpty = true` if no tests but `thresholds.break` is set.

```text
19:04:39 (3507) INFO ProjectReader Found 1 of 23 file(s) to be mutated.
19:04:40 (3507) INFO Instrumenter Instrumented 1 source file(s) with 58 mutant(s)
19:04:40 (3507) INFO ConcurrencyTokenProvider Creating 1 checker process(es) and 1 test runner process(es).
19:04:45 (3507) INFO Sandbox Running build command "tsc --build" in "/workspaces/graphql-markdown/packages/docusaurus/.stryker-tmp/sandbox6232593".
19:04:48 (3507) INFO DryRunExecutor Starting initial test run (jest test runner with "perTest" coverage analysis). This may take a while.
19:04:48 (3507) INFO DryRunExecutor No tests were found
19:05:49 (3507) INFO DashboardReporter The report was not send to the dashboard. The dashboard.project and/or dashboard.version values were missing and not detected to be running on a build server.
19:05:49 (3507) ERROR MutationTestReportHelper Final mutation score 0.00 under breaking threshold 70, setting exit code to 1 (failure).
19:05:49 (3507) INFO MutationTestReportHelper (improve mutation score or
```

If no tests found and `allowEmpty = true`, then StrykerJS should exist without error (code 0) and ignore the breaking threshold.

**Stryker config**

```json
{
  "buildCommand": "tsc --build",
  "checkers": ["typescript"],
  "commandRunner": { "command:" "npm run test:ci" },
  "coverageAnalysis": "perTest",
  "dashboard": {},
  "ignorePatterns": ["assets", "build", "config", "coverage", "dist", "docs"],
  "ignoreStatic": true,
  "inPlace": false,
  "jest": {
    "config": {
      "bail": false,
      "collectCoverage": false,
      "collectCoverageFrom": ["<rootDir>/src/**/*.ts"],
      "moduleFileExtensions": ["js", "ts"],
      "notify": false,
      "preset": "ts-jest",
      "reporters": [],
      "roots": [
        "<rootDir>/src/",
        "<rootDir>/tests/",
        "<rootDir>/tests/__mocks__",
      ],
      "testEnvironment": "node",
      "testMatch": ["<rootDir>/tests/unit/**/(*.)+(spec|test).ts"],
      "transform": {
        "^.+\\.ts$": [
          "ts-jest",
          {
            "tsconfig": "<rootDir>/tsconfig.test.json",
          },
        ],
      },
      "verbose": false,
    },
    "enableFindRelatedTests": true,
    "projectType": "custom",
  },
  "mutate": ["src/**/*.ts", "!src/**/prettier.ts"],
  "packageManager": "npm",
  "reporters": ["html"],
  "symlinkNodeModules": true,
  "testRunner": "jest",
  "testRunnerNodeArgs": ["--experimental-vm-modules"],
  "thresholds": { "high": 85, "low": 75, "break": 70 },
  "tsconfigFile": "tsconfig.json",
  "typescriptChecker": {
    "prioritizePerformanceOverAccuracy": true,
  },
  "warnings": true,
}
```

**Test runner config**

If relevant, see https://github.com/graphql-markdown/graphql-markdown/blob/main/config/jest.config.js

**Stryker environment**

```
+-- @stryker-mutator/core@7.2.0
+-- @stryker-mutator/jest-runner@7.2.0
+-- @stryker-mutator/typescript-checker@7.2.0
```

```
+-- jest@29.7.0
+-- ts-jest@29.1.1
```

**Test runner environment**

```shell
npm run stryker -w @graphql-markdown/docusaurus -- --allowEmpty --reporters dashboard --dashboard.module docusaurus
```

**Your Environment**

| software         | version(s)
| ---------------- | -------
| node             | 18.18.0
| npm              | 9.5.1
| Operating System | Linux 5.15.0-1041-azure x86_64

**Add stryker.log**

[stryker.log](https://github.com/stryker-mutator/stryker-js/files/12849584/stryker.log)

### PR DESCRIPTION
**Pull request by edno**
This should fix #4475 by short circuiting the test executor when no tests found and `allowEmpty=true` (same behaviour as `dryRunOnly`):
- prevent thresholds evaluation 
- avoid unnecessary run of the mutation executor (save some processing time)

### CODE REVIEW

File: packages/core/test/unit/process/4-mutation-test-executor.spec.ts [hunk 1]
```diff
@@ -401,4 +405,22 @@ describe(MutationTestExecutor.name, () => {
     expect(checker.check).not.called;
     expect(actualResults).empty;
   });
+
+  it('should short circuit when no tests found and allowEmpty is enabled', async () => {
```
Comment (nicojs) [inline packages/core/test/unit/process/4-mutation-test-executor.spec.ts:409]: Please also add a test to not short circuit when no tests are found and `allowEmpty` is disabled.

**Review by nicojs — CHANGES_REQUESTED** [review]
Great addition. I also like the unit test you've added, could you also add one more (see my remark).

File: packages/core/test/unit/process/4-mutation-test-executor.spec.ts (hunk 1 above)
Comment (edno) [reply packages/core/test/unit/process/4-mutation-test-executor.spec.ts:409]: Hi @nicojs 
I added a test when short circuit should be skipped if no tests are found and allowEmpty is disabled.

**Comment by nicojs** [comment]
Awesome, Thanks!

**Pull request merged**